File: apps/studio-web/src/api/storyboard.ts
/**
 * Storyboard API client — TypeScript types and fetch functions
 * for the unified paragraph-based pipeline.
 *
 * Mirrors backend models in creator_domain.models.storyboard.
 */

const API_BASE = "/api/creator";

// --------------- types ---------------

export type ParagraphStatus =
  | "idle"
  | "generating_image"
  | "generating_audio"
  | "generating_subtitles"
  | "ready"
  | "stale"
  | "failed";

export interface StaleFlags {
  prompt_stale: boolean;
  image_stale: boolean;
  audio_stale: boolean;
  subtitles_stale: boolean;
}

export interface SubtitleEntry {
  index: number;
  start: string;
  end: string;
  text: string;
}

export interface StoryboardParagraph {
  section_id: string;
  order: number;
  text: string;
  display_text: string | null;
  image_prompt: string | null;
  image_url: string | null;
  audio_url: string | null;
  audio_duration: number | null;
  subtitles_url: string | null;
  subtitle_entries: SubtitleEntry[] | null;
  status: ParagraphStatus;
  stale_flags: StaleFlags | null;
  scene_id: string | null;
  image_asset_id: number | null;
  audio_artifact_id: number | null;
  subtitle_artifact_id: number | null;
}

export interface StoryboardResponse {
  run_id: number;
  paragraphs: StoryboardParagraph[];
  render_ready: boolean;
  total_paragraphs: number;
  ready_paragraphs: number;
}

export interface ParagraphAudioParams {
  tts_model?: string;
  voice?: string;
}

export interface ParagraphSubtitlesParams {
  subtitle_model?: string;
  subtitle_format?: string;
}

// --------------- API functions ---------------

/**
 * Fetch full storyboard for a run — assembles paragraphs from
 * script sections + visual plan + images + audio + subtitles.
 */
export async function fetchStoryboard(runId: number): Promise<StoryboardResponse> {
  const res = await fetch(`${API_BASE}/runs/${runId}/storyboard`);
  if (!res.ok) {
    const body = await res.json().catch(() => null);
    throw new Error(body?.detail ?? `Failed to fetch storyboard (${res.status})`);
  }
  return res.json();
}

/**
 * Generate audio for a single paragraph (section).
 */
export async function generateParagraphAudio(
  runId: number,
  sectionId: string,
  params: ParagraphAudioParams = {},
): Promise<{ task_id: string }> {
  const res = await fetch(
    `${API_BASE}/runs/${runId}/storyboard/paragraphs/${sectionId}/generate-audio`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        tts_model: params.tts_model ?? "qwen3-tts",
        voice: params.voice ?? "default",
      }),
    },
  );
  if (!res.ok) {
    const body = await res.json().catch(() => null);
    throw new Error(body?.detail ?? `Generate audio failed (${res.status})`);
  }
  return res.json();
}

/**
 * Generate subtitles for a single paragraph (section).
 * Requires audio to already exist for this paragraph.
 */
export async function generateParagraphSubtitles(
  runId: number,
  sectionId: string,
  params: ParagraphSubtitlesParams = {},
): Promise<{ task_id: string }> {
  const res = await fetch(
    `${API_BASE}/runs/${runId}/storyboard/paragraphs/${sectionId}/generate-subtitles`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        subtitle_model: params.subtitle_model ?? "whisper-small",
        subtitle_format: params.subtitle_format ?? "srt",
      }),
    },
  );
  if (!res.ok) {
    const body = await res.json().catch(() => null);
    throw new Error(body?.detail ?? `Generate subtitles failed (${res.status})`);
  }
  return res.json();
}

/**
 * Generate audio for ALL paragraphs in bulk.
 */
export async function generateAllParagraphAudio(
  runId: number,
  params: ParagraphAudioParams = {},
): Promise<{ dispatched: number }> {
  const res = await fetch(
    `${API_BASE}/runs/${runId}/storyboard/generate-all-audio`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        tts_model: params.tts_model ?? "qwen3-tts",
        voice: params.voice ?? "default",
      }),
    },
  );
  if (!res.ok) {
    const body = await res.json().catch(() => null);
    throw new Error(body?.detail ?? `Bulk audio generation failed (${res.status})`);
  }
  return res.json();
}

/**
 * Generate subtitles for ALL paragraphs that have audio.
 */
export async function generateAllParagraphSubtitles(
  runId: number,
  params: ParagraphSubtitlesParams = {},
): Promise<{ dispatched: number }> {
  const res = await fetch(
    `${API_BASE}/runs/${runId}/storyboard/generate-all-subtitles`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        subtitle_model: params.subtitle_model ?? "whisper-small",
        subtitle_format: params.subtitle_format ?? "srt",
      }),
    },
  );
  if (!res.ok) {
    const body = await res.json().catch(() => null);
    throw new Error(body?.detail ?? `Bulk subtitle generation failed (${res.status})`);
  }
  return res.json();
}

File: packages/creator-domain/creator_domain/models/storyboard.py
"""Storyboard domain models for unified paragraph-based pipeline.

Each storyboard paragraph represents one section of the script, with its
associated image, audio, and subtitles.  The StoryboardResponse aggregates
all paragraphs for a run into a single view.

Key invariant: 1 paragraph = 1 image = 1 audio = N subtitles.
"""

from __future__ import annotations

from enum import Enum
from typing import ClassVar

from pydantic import BaseModel, ConfigDict, Field


class ParagraphStatus(str, Enum):
    """Per-paragraph generation status."""

    IDLE = "idle"
    GENERATING_IMAGE = "generating_image"
    GENERATING_AUDIO = "generating_audio"
    GENERATING_SUBTITLES = "generating_subtitles"
    READY = "ready"
    STALE = "stale"
    FAILED = "failed"


class StaleFlags(BaseModel):
    model_config: ClassVar[ConfigDict] = ConfigDict(extra="forbid")

    """Tracks which downstream artifacts are stale after an upstream edit."""

    prompt_stale: bool = False
    image_stale: bool = False
    audio_stale: bool = False
    subtitles_stale: bool = False


class StoryboardParagraph(BaseModel):
    """One paragraph in the unified storyboard view."""

    model_config: ClassVar[ConfigDict] = ConfigDict(extra="forbid")

    section_id: str = Field(max_length=128)
    order: int = Field(ge=0)
    text: str
    display_text: str | None = None
    image_prompt: str | None = None
    image_url: str | None = None
    audio_url: str | None = None
    audio_duration: float | None = Field(default=None, ge=0)
    subtitles_url: str | None = None
    subtitle_entries: list[dict[str, object]] | None = None
    status: ParagraphStatus = ParagraphStatus.IDLE
    stale_flags: StaleFlags | None = None

    # Scene/asset metadata for cross-reference
    scene_id: str | None = Field(default=None, max_length=128)
    image_asset_id: int | None = Field(default=None, ge=1)
    audio_artifact_id: int | None = Field(default=None, ge=1)
    subtitle_artifact_id: int | None = Field(default=None, ge=1)


class StoryboardResponse(BaseModel):
    """Full storyboard data for a run."""

    model_config: ClassVar[ConfigDict] = ConfigDict(extra="forbid")

    run_id: int = Field(ge=1)
    paragraphs: list[StoryboardParagraph]
    render_ready: bool = False
    total_paragraphs: int = Field(ge=0, default=0)
    ready_paragraphs: int = Field(ge=0, default=0)
